File: backend/src/main/java/com/spooltracker/entity/Spool.java
package com.spooltracker.entity;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import io.quarkus.hibernate.orm.panache.PanacheEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;

@Entity
@Table(name = "spool")
public class Spool extends PanacheEntity {

    @Column(nullable = false, unique = true, updatable = false)
    public String uid;

    // Spool physical type (Plastic, Refill, Cardboard)
    @Enumerated(EnumType.STRING)
    @Column(name = "spool_type")
    public SpoolType spoolType = SpoolType.PLASTIC;

    @NotNull(message = "Filament type is required")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "filament_type_id", nullable = false)
    public FilamentType filamentType;

    @NotNull(message = "Color is required")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "color_id", nullable = false)
    public FilamentColor color;

    @NotNull(message = "Manufacturer is required")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "manufacturer_id", nullable = false)
    public Manufacturer manufacturer;

    // Legacy location enum (kept for backward compatibility during migration)
    @Enumerated(EnumType.STRING)
    @Column(name = "location")
    public SpoolLocation legacyLocation;

    // New location reference
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "storage_location_id")
    public Location storageLocation;

    // Location details (e.g., "AMS Slot 1", "Rack A-3", "Printer 1")
    public String locationDetails;

    // Helper method to get location name (works with both old and new system)
    public String getLocationName() {
        if (storageLocation != null) {
            return storageLocation.name;
        }
        return legacyLocation != null ? legacyLocation.getDisplayName() : null;
    }

    // Helper method to get location type
    public String getLocationType() {
        if (storageLocation != null) {
            return storageLocation.locationType;
        }
        return legacyLocation != null ? legacyLocation.name() : null;
    }

    // Weight tracking
    @Positive(message = "Initial weight must be positive")
    public Double initialWeightGrams;

    @Positive(message = "Current weight must be positive")
    public Double currentWeightGrams;

    // Spool metadata
    public LocalDate purchaseDate;
    public LocalDate openedDate;
    public LocalDate lastUsedDate;

    // Price tracking
    public Double purchasePrice;
    public String purchaseCurrency;

    // Notes
    @Column(length = 1000)
    public String notes;

    // Custom color number for customer ordering (e.g., "5" for "item number 5")
    @Column(name = "color_number")
    public String colorNumber;

    // Timestamps
    @Column(nullable = false, updatable = false)
    public LocalDateTime createdAt;

    @Column(nullable = false)
    public LocalDateTime updatedAt;

    // Whether this spool is empty/finished
    public Boolean isEmpty = false;

    @PrePersist
    public void prePersist() {
        if (uid == null) {
            uid = UUID.randomUUID().toString();
        }
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public static Spool findByUid(String uid) {
        return find("uid", uid).firstResult();
    }

    public static List<Spool> findByLegacyLocation(SpoolLocation location) {
        return list("legacyLocation", location);
    }

    public static List<Spool> findByStorageLocation(Long locationId) {
        return list("storageLocation.id", locationId);
    }

    public static List<Spool> findByFilamentType(Long filamentTypeId) {
        return list("filamentType.id", filamentTypeId);
    }

    public static List<Spool> findByManufacturer(Long manufacturerId) {
        return list("manufacturer.id", manufacturerId);
    }

    public static List<Spool> findByColor(Long colorId) {
        return list("color.id", colorId);
    }

    public static List<Spool> findActive() {
        return list("isEmpty", false);
    }

    public static List<Spool> findEmpty() {
        return list("isEmpty", true);
    }

    public static List<Spool> findByColorNumber(String colorNumber) {
        return list("colorNumber", colorNumber);
    }

    public Double getRemainingPercentage() {
        if (initialWeightGrams == null || initialWeightGrams == 0 || currentWeightGrams == null) {
            return null;
        }
        return (currentWeightGrams / initialWeightGrams) * 100;
    }
}

File: frontend/src/schemas/spool.ts
import { z } from 'zod';

export const createSpoolSchema = z.object({
  filamentTypeId: z.number({ message: 'Filament type is required' }).min(1, 'Filament type is required'),
  colorId: z.number({ message: 'Color is required' }).min(1, 'Color is required'),
  manufacturerId: z.number().min(1, 'Manufacturer is required').optional(),
  spoolType: z.enum(['PLASTIC', 'REFILL', 'CARDBOARD']).optional(),
  storageLocationId: z.number().optional(),
  location: z.enum(['AMS', 'PRINTER', 'RACK', 'STORAGE', 'IN_USE', 'EMPTY']).optional(),
  locationDetails: z.string().max(500, 'Location details must be less than 500 characters').optional(),
  initialWeightGrams: z.union([
    z.number().positive('Initial weight must be positive'),
    z.string().transform((val) => {
      if (val === '') return undefined;
      const num = parseFloat(val);
      return isNaN(num) ? undefined : num;
    }).pipe(z.number().positive('Initial weight must be positive').optional()),
    z.undefined(),
    z.null().transform(() => undefined),
  ]).optional(),
  currentWeightGrams: z.union([
    z.number().positive('Current weight must be positive'),
    z.string().transform((val) => {
      if (val === '') return undefined;
      const num = parseFloat(val);
      return isNaN(num) ? undefined : num;
    }).pipe(z.number().positive('Current weight must be positive').optional()),
    z.undefined(),
    z.null().transform(() => undefined),
  ]).optional(),
  purchaseDate: z.string().optional(),
  openedDate: z.string().optional(),
  lastUsedDate: z.string().optional(),
  purchasePrice: z.union([
    z.number().min(0, 'Price cannot be negative'),
    z.string().transform((val) => {
      if (val === '') return undefined;
      const num = parseFloat(val);
      return isNaN(num) ? undefined : num;
    }).pipe(z.number().min(0, 'Price cannot be negative').optional()),
    z.undefined(),
    z.null().transform(() => undefined),
  ]).optional(),
  purchaseCurrency: z.string()
    .max(10, 'Currency code must be 10 characters or less')
    .refine(
      (val) => !val || /^[A-Z]{3}$/.test(val),
      { message: 'Currency code must be a 3-letter ISO 4217 code (e.g., USD, EUR, BRL)' }
    )
    .optional(),
  colorNumber: z.string().max(50, 'Color number must be 50 characters or less').optional(),
  notes: z.string().max(1000, 'Notes must be less than 1000 characters').optional(),
}).refine(
  (data) => {
    // At least one location must be provided
    return data.storageLocationId || data.location;
  },
  {
    message: 'Either storage location or legacy location must be provided',
    path: ['location'],
  }
).refine(
  (data) => {
    // If currentWeightGrams is provided, it should not exceed initialWeightGrams
    if (data.currentWeightGrams && data.initialWeightGrams) {
      return data.currentWeightGrams <= data.initialWeightGrams;
    }
    return true;
  },
  {
    message: 'Current weight cannot exceed initial weight',
    path: ['currentWeightGrams'],
  }
);

export const updateSpoolSchema = createSpoolSchema.partial();

export type CreateSpoolFormData = z.infer<typeof createSpoolSchema>;
export type UpdateSpoolFormData = z.infer<typeof updateSpoolSchema>;

